File: src/nuclearbot/gui/components/ModeratorPanel.java
package nuclearbot.gui.components;

import nuclearbot.client.Moderators;
import nuclearbot.gui.NuclearBotGUI;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionListener;

/*
 * Copyright (C) 2017 NuclearCoder
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * GUI panel for the moderator list.<br>
 * <br>
 * NuclearBot (https://github.com/NuclearCoder/nuclear-bot/)<br>
 *
 * @author NuclearCoder (contact on the GitHub repo)
 */
public class ModeratorPanel extends JPanel {

    private final NuclearBotGUI m_gui;

    private final DefaultListModel<String> m_moderators;

    public ModeratorPanel(final NuclearBotGUI gui)
    {
        super(new BorderLayout());

        m_gui = gui;

        m_moderators = new DefaultListModel<>();

        final JList<String> list = new JList<>(m_moderators);
        final JScrollPane listScroll = new JScrollPane(list);

        listScroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        listScroll.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_AS_NEEDED);

        final JPanel side = new JPanel(new FlowLayout());
        {
            final JTextField nameField = new JTextField(10);
            final JButton addButton = new JButton("Add");
            final JButton removeButton = new JButton("Remove");

            final ActionListener addAction = e ->
            {
                final String name = nameField.getText().toLowerCase().trim();
                if (!name.isEmpty())
                {
                    nameField.setText("");
                    addModerator(name);
                }
            };

            nameField.addActionListener(addAction);
            addButton.addActionListener(addAction);

            removeButton.addActionListener(e ->
                    list.getSelectedValuesList().forEach(this::removeModerator));

            side.add(nameField);
            side.add(addButton);
            side.add(removeButton);
        }

        add(listScroll, BorderLayout.CENTER);
        add(side, BorderLayout.SOUTH);

        // init
        Moderators.getModerators().forEach(m_moderators::addElement);
    }

    public void addModerator(final String name)
    {
        if (Moderators.isModerator(name))
        {
            m_gui.getDialogs().warning("User \"" + name + "\" is already moderator.", "Already moderator");
        }
        else
        {
            Moderators.addModerator(name);

            // this is to ensure the list remains sorted
            if (m_moderators.isEmpty())
            {
                m_moderators.addElement(name);
            }
            else
            {
                int index = 0;
                String element;
                do
                {
                    element = m_moderators.get(index);
                    if (element.compareTo(name) > 0) // insert the element at the right index
                    {
                        m_moderators.add(index, name);
                        break;
                    }
                    index++;
                }
                while (index < m_moderators.size());
            }
        }
    }

    public void removeModerator(final String name)
    {
        if (!Moderators.isModerator(name))
        {
            m_gui.getDialogs().warning("User \"" + name + "\" is not moderator.", "Not moderator");
        }
        else
        {
            Moderators.removeModerator(name);
            m_moderators.removeElement(name);
        }
    }

}

File: src/nuclearbot/gui/components/ModeratorPanel.kt
package nuclearbot.gui.components

import nuclearbot.client.Moderators
import nuclearbot.gui.NuclearBotGUI
import java.awt.BorderLayout
import java.awt.FlowLayout
import java.awt.event.ActionEvent
import javax.swing.*
import javax.swing.border.CompoundBorder
import javax.swing.border.EmptyBorder

/*
 * Copyright (C) 2017 NuclearCoder
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * GUI panel for the moderator list.<br></br>
 * <br></br>
 * NuclearBot (https://github.com/NuclearCoder/nuclear-bot/)<br></br>

 * @author NuclearCoder (contact on the GitHub repo)
 */
class ModeratorPanel(private val gui: NuclearBotGUI) : JPanel(BorderLayout()) {

    private val moderators = DefaultListModel<String>()

    init {
        val list = JList(moderators).apply {
            layoutOrientation = JList.VERTICAL_WRAP
            setCellRenderer(object : DefaultListCellRenderer() {
                override fun getListCellRendererComponent(list: JList<*>, value: Any?,
                                                          index: Int, isSelected: Boolean,
                                                          cellHasFocus: Boolean) =
                        (super.getListCellRendererComponent(list, value, index,
                                isSelected, cellHasFocus) as JLabel).apply {
                            val matteBorder = BorderFactory.createMatteBorder(1, 1, 1, 1,
                                    UIManager.getColor("Table.gridColor"))
                            val margin = EmptyBorder(0, 5, 0, 5)
                            border = CompoundBorder(matteBorder, margin)
                            horizontalAlignment = JLabel.CENTER
                            verticalAlignment = JLabel.CENTER
                        }
            })
        }

        val listScroll = JScrollPane(list).apply {
            horizontalScrollBarPolicy = JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED
            verticalScrollBarPolicy = JScrollPane.VERTICAL_SCROLLBAR_AS_NEEDED
        }

        val side = JPanel(FlowLayout()).apply {
            val nameField = JTextField(20)
            val addButton = JButton("Add")
            val removeButton = JButton("Remove").apply {
                addActionListener {
                    list.selectedValuesList.forEach(this@ModeratorPanel::removeModerator)
                }
            }

            val addAction = { _: ActionEvent ->
                val name = nameField.text.toLowerCase().trim { it <= ' ' }
                if (!name.isEmpty()) {
                    nameField.text = ""
                    addModerator(name)
                }
            }

            nameField.addActionListener(addAction)
            addButton.addActionListener(addAction)

            add(nameField)
            add(addButton)
            add(removeButton)
        }

        add(listScroll, BorderLayout.CENTER)
        add(side, BorderLayout.SOUTH)

        // init
        Moderators.getModerators().forEach(moderators::addElement)
    }

    fun addModerator(name: String) {
        if (Moderators.isModerator(name)) {
            gui.dialogs.warning("User \"$name\" is already moderator.", "Already moderator")
        } else {
            Moderators.addModerator(name)

            // this is to ensure the list remains sorted
            if (moderators.isEmpty) {
                moderators.addElement(name)
            } else {
                moderators.add(moderators.elements().asSequence().indexOfLast { it < name }, name)
            }
        }
    }

    fun removeModerator(name: String) {
        if (!Moderators.isModerator(name)) {
            gui.dialogs.warning("User \"$name\" is not moderator.", "Not moderator")
        } else {
            Moderators.removeModerator(name)
            moderators.removeElement(name)
        }
    }

}
